Library/Homebrew/test/os/mac/keg_spec.rb

```diff
@@ -5,28 +5,65 @@
 
   subject { described_class.new(keg_path) }
 
-  describe "#mach_o_files" do
-    let(:keg_path) { HOMEBREW_CELLAR/"a/1.0" }
+  let(:keg_path) { HOMEBREW_CELLAR/"a/1.0" }
+  let(:file) { keg_path/"lib/i386.dylib" }
 
-    before(:each) { (keg_path/"lib").mkpath }
-
-    after(:each) { subject.unlink }
+  before(:each) do
+    (keg_path/"lib").mkpath
+    cp dylib_path("i386"), file
+  end
 
-    it "skips hardlinks" do
-      cp dylib_path("i386"), keg_path/"lib/i386.dylib"
-      ln keg_path/"lib/i386.dylib", keg_path/"lib/i386_hardlink.dylib"
+  after(:each) { subject.unlink }
 
+  describe "#change_dylib_id" do
+    before(:each) do
       subject.link
-      expect(subject.mach_o_files.count).to eq(1)
     end
 
-    it "isn't confused by symlinks" do
-      cp dylib_path("i386"), keg_path/"lib/i386.dylib"
-      ln keg_path/"lib/i386.dylib", keg_path/"lib/i386_hardlink.dylib"
-      ln_s keg_path/"lib/i386.dylib", keg_path/"lib/i386_symlink.dylib"
+    it "does nothing if given id is same as file's dylib id" do
+      id = file.dylib_id
+      subject.change_dylib_id(id, file)
+      expect(file.dylib_id).to eq(id)
+    end
+
+    it "changes file's dylib id to given id" do
+      id = "blah.B.dylib"
+      file.ensure_writable do
+        subject.change_dylib_id(id, file)
+      end
+      expect(file.dylib_id).to eq(id)
+    end
+  end
 
+  describe "#change_install_name" do
+    before(:each) do
       subject.link
-      expect(subject.mach_o_files.count).to eq(1)
+    end
+
+    it "does nothing if given name is same as file's install name" do
+      initial_name = ""
+      file.ensure_writable do
+        subject.each_install_name_for(file) do |name|
+          initial_name = name
+          subject.change_install_name(name, name, file)
+        end
+      end
+      subject.each_install_name_for(file) do |name|
+        expect(name).to eq(initial_name)
+      end
+    end
+
+    it "changes file's install name to given name" do
+      new_name = ""
+      file.ensure_writable do
+        subject.each_install_name_for(file) do |name|
+          new_name = subject.fixed_name(file, name)
+          subject.change_install_name(name, new_name, file)
+        end
+      end
+      subject.each_install_name_for(file) do |name|
+        expect(name).to eq(new_name)
```

Comment:
Why are we expecting each install name in the file to be the same as the last fixed name (`new_name`)?

In the file we're testing against, the only install name is `/usr/lib/libSystem.B.dylib`. Since that begins with a `/`, it's one of the ones that we don't called `fixed_name` on in `fix_dynamic_linkage`:

```ruby
          # Don't fix absolute paths unless they are rooted in the build directory
          next if bad_name.start_with?("/") &&
                  !bad_name.start_with?(HOMEBREW_TEMP.to_s) &&
                  !bad_name.start_with?(HOMEBREW_TEMP.realpath.to_s)
```

Reply:
We are expecting each install name in the file to be same as the last fixed name because there is only one file and only one install name for the file is /usr/lib/libSystem.B.dylib.
I'll try thinking of a case for which we actually modify the install name.

Reply:
Got it. I think that we can invert this test, then, since we actually expect install names that begin with a `/` to *never* be rewritten. `expect(name).to_not eq(new_name)` should do the trick, if I'm remembering my RSpec correctly.

We might also want to amend this test to only attempt to rewrite the first install name in the binary, since we might update the fixture in the future to contain additional `LC_LOAD_*` commands.

Reply:
to rewrite the first install name in the binary, I should be running the loop only once. Is that what you mean, @woodruffw ?